File: src/application/products/product-service.ts
import { repositories } from '@/data/repositories'
import type { Product } from '@/domain/products/product'
import type { ProductDraft } from '@/domain/products/product-draft'
import type { ProductVersion } from '@/domain/products/product-version'
import type { ServingUnit } from '@/domain/products/serving-unit'
import type { ProductRepository } from '@/domain/repositories/product-repository'
import { createUuid } from '@/shared/utils/create-uuid'

export interface ProductListItem {
  product: Product
  currentVersion: ProductVersion
}

export interface ProductDetails extends ProductListItem {
  versions: ProductVersion[]
}

export class ProductService {
  constructor(private readonly productRepository: ProductRepository) {}

  async search(query = ''): Promise<ProductListItem[]> {
    const normalizedQuery = query.trim().toLocaleLowerCase()
    const products = await this.productRepository.getActive()
    const matchedProducts = products.filter((product) => {
      if (normalizedQuery.length === 0) {
        return true
      }

      return product.name.toLocaleLowerCase().includes(normalizedQuery)
        || product.barcode?.includes(query.trim()) === true
    })

    const items = await Promise.all(
      matchedProducts.map(async (product) => {
        const currentVersion = await this.productRepository.getVersionById(product.currentVersionId)

        return currentVersion === undefined ? undefined : { product, currentVersion }
      }),
    )

    return items.filter((item): item is ProductListItem => item !== undefined)
  }

  async getById(id: string): Promise<ProductDetails | undefined> {
    const product = await this.productRepository.getById(id)

    if (product === undefined || product.deletedAt !== undefined) {
      return undefined
    }

    const [currentVersion, versions] = await Promise.all([
      this.productRepository.getVersionById(product.currentVersionId),
      this.productRepository.getVersions(product.id),
    ])

    if (currentVersion === undefined) {
      return undefined
    }

    return { product, currentVersion, versions }
  }

  async create(draft: ProductDraft): Promise<ProductDetails> {
    const now = new Date().toISOString()
    const productId = createUuid()
    const version = createVersion(productId, 1, draft, now)
    const product: Product = {
      id: productId,
      name: draft.name.trim(),
      barcode: normalizeBarcode(draft.barcode),
      currentVersionId: version.id,
      createdAt: now,
      updatedAt: now,
    }

    await this.productRepository.create(product, version)

    return { product, currentVersion: version, versions: [version] }
  }

  async update(id: string, draft: ProductDraft): Promise<ProductDetails> {
    const details = await this.getById(id)

    if (details === undefined) {
      throw new ProductNotFoundError()
    }

    const now = new Date().toISOString()
    const updatedProduct: Product = {
      ...details.product,
      name: draft.name.trim(),
      barcode: normalizeBarcode(draft.barcode),
      updatedAt: now,
    }

    if (!hasVersionChanged(details.currentVersion, draft)) {
      await this.productRepository.update(updatedProduct)

      return { ...details, product: updatedProduct }
    }

    const version = createVersion(updatedProduct.id, details.currentVersion.versionNumber + 1, draft, now)
    updatedProduct.currentVersionId = version.id

    await this.productRepository.addVersionAndUpdateProduct(updatedProduct, version)

    return {
      product: updatedProduct,
      currentVersion: version,
      versions: [...details.versions, version],
    }
  }
}

export class ProductNotFoundError extends Error {
  constructor() {
    super('Product not found.')
    this.name = 'ProductNotFoundError'
  }
}

export const productService = new ProductService(repositories.products)

function createVersion(productId: string, versionNumber: number, draft: ProductDraft, createdAt: string): ProductVersion {
  return {
    id: createUuid(),
    productId,
    versionNumber,
    baseUnitType: draft.baseUnitType,
    baseAmount: draft.baseAmount,
    calories: draft.calories,
    protein: draft.protein,
    fat: draft.fat,
    carbs: draft.carbs,
    servingUnits: draft.servingUnits.map(createServingUnit),
    createdAt,
  }
}

function createServingUnit(draft: ProductDraft['servingUnits'][number]): ServingUnit {
  return {
    id: createUuid(),
    name: draft.name.trim(),
    conversionAmount: draft.conversionAmount,
    conversionUnit: draft.conversionUnit,
  }
}

function hasVersionChanged(version: ProductVersion, draft: ProductDraft): boolean {
  return version.baseUnitType !== draft.baseUnitType
    || version.baseAmount !== draft.baseAmount
    || version.calories !== draft.calories
    || version.protein !== draft.protein
    || version.fat !== draft.fat
    || version.carbs !== draft.carbs
    || !areServingUnitsEqual(version.servingUnits, draft.servingUnits)
}

function areServingUnitsEqual(
  savedUnits: ServingUnit[],
  draftUnits: ProductDraft['servingUnits'],
): boolean {
  return savedUnits.length === draftUnits.length && savedUnits.every((unit, index) => {
    const draft = draftUnits[index]

    return unit.name === draft.name.trim()
      && unit.conversionAmount === draft.conversionAmount
      && unit.conversionUnit === draft.conversionUnit
  })
}

function normalizeBarcode(barcode: string | undefined): string | undefined {
  const normalizedBarcode = barcode?.trim()

  return normalizedBarcode === '' ? undefined : normalizedBarcode
}

File: CaloriesTrackerIOS/Application/Products/ProductService.swift
import Foundation

struct ProductListItem: Identifiable, Hashable, Sendable {
    let product: Product
    let currentVersion: ProductVersion

    var id: UUID {
        product.id
    }
}

struct ProductDetails: Hashable, Sendable {
    let product: Product
    let currentVersion: ProductVersion
}

@MainActor
final class ProductService {
    private let repository: any ProductRepository

    init(repository: any ProductRepository) {
        self.repository = repository
    }

    func products(matching query: String) async throws -> [ProductListItem] {
        let products = try await repository.activeProducts(matching: query)
        var items: [ProductListItem] = []

        for product in products {
            guard let currentVersion = try await repository.version(id: product.currentVersionID) else {
                throw ProductServiceError.currentVersionNotFound
            }
            items.append(ProductListItem(product: product, currentVersion: currentVersion))
        }

        return items
    }

    func details(id: UUID) async throws -> ProductDetails? {
        guard let product = try await repository.product(id: id, includingDeleted: false) else {
            return nil
        }
        guard let currentVersion = try await repository.version(id: product.currentVersionID) else {
            throw ProductServiceError.currentVersionNotFound
        }

        return ProductDetails(product: product, currentVersion: currentVersion)
    }

    func versions(for productID: UUID) async throws -> [ProductVersion] {
        let versions = try await repository.versions(for: productID)
        return versions.sorted {
            if $0.versionNumber == $1.versionNumber {
                return $0.createdAt > $1.createdAt
            }
            return $0.versionNumber > $1.versionNumber
        }
    }

    @discardableResult
    func create(draft: ProductDraft) async throws -> UUID {
        let draft = try await validatedDraft(from: draft, excludingProductID: nil)
        let now = Date()
        let productID = UUID()
        let versionID = UUID()
        let version = ProductVersion(
            id: versionID,
            productID: productID,
            basedOnVersionID: nil,
            versionNumber: 1,
            baseUnit: draft.baseUnit,
            baseAmount: draft.baseAmount,
            nutrition: draft.nutrition,
            servingUnits: [],
            createdAt: now,
        )
        let product = Product(
            id: productID,
            name: draft.name,
            barcode: draft.barcode,
            currentVersionID: versionID,
            createdAt: now,
            updatedAt: now,
            deletedAt: nil,
        )

        try await repository.create(product, initialVersion: version)
        return productID
    }

    func update(productID: UUID, draft: ProductDraft) async throws {
        guard let details = try await details(id: productID) else {
            throw ProductServiceError.productNotFound
        }

        let draft = try await validatedDraft(from: draft, excludingProductID: productID)
        let versionedFieldsChanged = details.currentVersion.baseUnit != draft.baseUnit
            || details.currentVersion.baseAmount != draft.baseAmount
            || details.currentVersion.nutrition != draft.nutrition
        let metadataChanged = details.product.name != draft.name
            || details.product.barcode != draft.barcode

        guard versionedFieldsChanged || metadataChanged else {
            return
        }

        let now = Date()

        if versionedFieldsChanged {
            let versionID = UUID()
            let servingUnits = details.currentVersion.servingUnits.map { oldUnit in
                ServingUnit(
                    id: UUID(),
                    productVersionID: versionID,
                    position: oldUnit.position,
                    name: oldUnit.name,
                    conversionAmount: oldUnit.conversionAmount,
                    conversionUnit: oldUnit.conversionUnit,
                )
            }
            let nextVersion = ProductVersion(
                id: versionID,
                productID: details.product.id,
                basedOnVersionID: details.currentVersion.id,
                versionNumber: details.currentVersion.versionNumber + 1,
                baseUnit: draft.baseUnit,
                baseAmount: draft.baseAmount,
                nutrition: draft.nutrition,
                servingUnits: servingUnits,
                createdAt: now,
            )
            let updatedProduct = Product(
                id: details.product.id,
                name: draft.name,
                barcode: draft.barcode,
                currentVersionID: nextVersion.id,
                createdAt: details.product.createdAt,
                updatedAt: now,
                deletedAt: details.product.deletedAt,
            )

            try await repository.append(nextVersion, settingCurrentVersionOf: updatedProduct)
        } else {
            let updatedProduct = Product(
                id: details.product.id,
                name: draft.name,
                barcode: draft.barcode,
                currentVersionID: details.product.currentVersionID,
                createdAt: details.product.createdAt,
                updatedAt: now,
                deletedAt: details.product.deletedAt,
            )
            try await repository.saveLogicalMetadata(updatedProduct)
        }
    }

    func softDelete(productID: UUID) async throws {
        guard try await repository.product(id: productID, includingDeleted: false) != nil else {
            throw ProductServiceError.productNotFound
        }

        try await repository.softDeleteProduct(id: productID, at: Date())
    }

    private func validatedDraft(
        from draft: ProductDraft,
        excludingProductID: UUID?,
    ) async throws -> ProductDraft {
        let name = draft.name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else {
            throw ProductServiceError.nameRequired
        }
        guard draft.baseAmount.isFinite, draft.baseAmount > 0 else {
            throw ProductServiceError.invalidBaseAmount
        }
        guard nutritionIsValid(draft.nutrition) else {
            throw ProductServiceError.invalidNutrition
        }

        let barcode = draft.barcode?.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedBarcode = barcode?.isEmpty == true ? nil : barcode
        if let normalizedBarcode,
           let existing = try await repository.product(withBarcode: normalizedBarcode),
           existing.id != excludingProductID {
            throw ProductServiceError.barcodeAlreadyInUse
        }

        return ProductDraft(
            name: name,
            barcode: normalizedBarcode,
            baseUnit: draft.baseUnit,
            baseAmount: draft.baseAmount,
            nutrition: draft.nutrition,
        )
    }

    private func nutritionIsValid(_ nutrition: Nutrition) -> Bool {
        [nutrition.calories, nutrition.protein, nutrition.fat, nutrition.carbs]
            .allSatisfy { $0.isFinite && $0 >= 0 }
    }
}

enum ProductServiceError: LocalizedError {
    case productNotFound
    case currentVersionNotFound
    case nameRequired
    case invalidBaseAmount
    case invalidNutrition
    case barcodeAlreadyInUse

    var errorDescription: String? {
        switch self {
        case .productNotFound:
            "Продукт не найден."
        case .currentVersionNotFound:
            "Не удалось найти текущую версию продукта."
        case .nameRequired:
            "Укажите название продукта."
        case .invalidBaseAmount:
            "Количество должно быть больше нуля."
        case .invalidNutrition:
            "Пищевая ценность должна содержать конечные неотрицательные значения."
        case .barcodeAlreadyInUse:
            "Этот штрихкод уже используется другим продуктом."
        }
    }
}
